### PR TITLE
Pin macro tweaks

### PR DESCRIPTION
**Pull request by MajorBreakfast**

### CODE REVIEW

File: futures-core/src/macros/pin.rs [hunk 1]
```diff
@@ -1,25 +1,35 @@
 /// A pinned projection of a struct field.
 ///
+/// To make using this macro safe, two things need to be ensured:
+/// - If [`Drop`] is implemented, [`drop`] must not move the field's value
+/// - If [`Unpin`] is implemented, it has to be conditionally:
+///   The type can only be [`Unpin`] if the field's type is [`Unpin`].
```
Comment (MajorBreakfast) [inline futures-core/src/macros/pin.rs]: This is the new explanation for the `unsafe_pinned` macro. Is that alright?

File: futures-core/src/macros/pin.rs [hunk 2]
```diff
@@ -34,9 +44,10 @@ macro_rules! unsafe_pinned {
 
 /// An unpinned projection of a struct field.
 ///
-/// This macro is unsafe because it returns a normal non-pin reference to
-/// the struct field. It is up to the programmer to ensure that the contained
-/// value is either not moved at all or only moved when it's safe.
+/// This macro is unsafe because it creates a method that returns a normal
+/// non-pin reference to the struct field. It is up to the programmer to ensure
+/// that the contained value can be considered not pinned in the current
+/// context.
```
Comment (MajorBreakfast) [inline futures-core/src/macros/pin.rs]: Explanation for the `unsafe_unpinned` macro

File: futures-core/src/macros/pin.rs [hunk 3]
```diff
@@ -48,7 +59,7 @@ macro_rules! unsafe_pinned {
 /// }
 ///
 /// impl Foo {
-///     unsafe_unpinned!(field -> Bar);
+///     unsafe_unpinned!(field: Bar);
```
Comment (MajorBreakfast) [inline futures-core/src/macros/pin.rs]: New syntax: Inspired by struct field def syntax. => Copy/paste possible

**Comment by cramertj** [comment]
This looks great, too, thanks! I'll merge once you've had a chance to rebase.

**Comment by MajorBreakfast** [comment]
Rebased

**Pull request merged**